pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -296,15 +298,301 @@ func (m *kubeGenericRuntimeManager) GetPods(all bool) ([]*kubecontainer.Pod, err
 	return result, nil
 }
 
-// SyncPod syncs the running pod into the desired pod.
-func (m *kubeGenericRuntimeManager) SyncPod(pod *api.Pod, _ api.PodStatus,
-	podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret,
-	backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
-	result.Fail(fmt.Errorf("not implemented"))
+// podContainerSpecChanges keeps information on changes that need to happen for a pod.
+type podContainerSpecChanges struct {
+	// Whether need to create a new sandbox.
+	CreateSandbox bool
+	// The id of existing sandbox if not needed to create a new one.
+	SandboxID string
+	// The attempt number of creating sandboxes for the pod.
+	Attempt uint32
+
+	ContainersToStart map[int]string
+	ContainersToKeep  map[kubecontainer.ContainerID]int
+}
+
+// podSandboxChanged checks whether the spec of the pod is changed.
+func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *api.Pod, podStatus *kubecontainer.PodStatus) (changed bool, attempt uint32, sandboxID string) {
+	if len(podStatus.SandboxStatuses) == 0 {
+		return true, 0, ""
+	}
+
+	readySandboxCount := 0
+	for _, s := range podStatus.SandboxStatuses {
+		if s.GetState() == runtimeApi.PodSandBoxState_READY {
+			readySandboxCount++
+		}
+	}
+
+	// Needs to create a new sandbox when readySandboxCount > 1 but the latest sandbox is not ready.
+	sandboxStatus := podStatus.SandboxStatuses[0]
+	if readySandboxCount > 1 || sandboxStatus.GetState() != runtimeApi.PodSandBoxState_READY {
```

Comment:
I think the comment `Needs to create a new sandbox when readySandboxCount > 1 but the latest sandbox is not ready.` will translate to `if readySandboxCount > 1 && sandboxStatus.GetState() != runtimeApi.PodSandBoxState_READY` ?